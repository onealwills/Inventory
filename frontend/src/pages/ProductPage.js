import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { detailsProduct } from "../actions/productActions";
import LoadingBox from "../components/LoadingBox";
import MessageBox from "../components/MessageBox";

export default function ProductPage(props) {
  const dispatch = useDispatch();
  const [qty, setQty] = useState(1);
  const productId = props.match.params.id;
  const productDetails = useSelector((state) => state.productDetails);
  const { loading, error, product } = productDetails;
  // console.log("productdetails >>>", productDetails);

  useEffect(() => {
    dispatch(detailsProduct(productId));
  }, [dispatch, productId]);

  const addToCartHandler = () => {
    props.history.push(`/cart/${productId}?qty=${qty}`);
  };
  return (
    <div className="productpage-container">
      {loading ? (
        <LoadingBox></LoadingBox>
      ) : error ? (
        <MessageBox type="danger">{error}</MessageBox>
      ) : (
        <div className="productpage">
          <div className="flex-2">
            <div>
              <img
                className="large-image"
                src={product.image}
                alt={product.model}
              />
            </div>
            <div>
              <img
                className="small-image"
                src={product.image}
                alt={product.model}
              />
              <img
                className="small-image"
                src={product.image}
                alt={product.model}
              />
              <img
                className="small-image"
                src={product.image}
                alt={product.model}
              />
            </div>
          </div>

          <div className="flex-1">
            <div>
              <h3>Product type: {product.type}</h3>
              <h2>
                <span>Car: </span>
                {product.make} {product.model} {product.year}
              </h2>

              <h2>Can Also fit:</h2>
              <div className="also-fit">Toyota venza 2019</div>
              <div className="status-qty">
                <h3>
                  Price : <span>&#8358;</span>
                  {product.price}
                </h3>
                <div className="status">
                  <h3>Status:</h3>
                  {product.stockQty > 0 ? (
                    <h3 className="success">In stock</h3>
                  ) : (
                    <h3 className="danger">Out Of Stock</h3>
                  )}
                </div>

                <div>
                  {product.stockQty > 0 && (
                    <div>
                      <div className="qty">
                        <h3>qty:</h3>
                        <select
                          className="select-qty"
                          value={qty}
                          onChange={(e) => setQty(e.target.value)}
                        >
                          {[...Array(product.stockQty).keys()].map((x) => (
                            <option key={x + 1} value={x + 1}>
                              {x + 1}
                            </option>
                          ))}
                        </select>
                      </div>
                      <button
                        onClick={addToCartHandler}
                        className="fw addtocart-btn"
                      >
                        Add to cart
                      </button>
                    </div>
                  )}
                </div>
              </div>
            </div>
          </div>
        </div>
      )}
    </div>
  );
}
