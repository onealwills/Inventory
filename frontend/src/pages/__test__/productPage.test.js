import React from "react";
import { shallow, configure } from "enzyme";
import ProductPage from "../ProductPage";
import Adapter from "enzyme-adapter-react-16";
import { useSelector } from "react-redux";

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

configure({ adapter: new Adapter() });
describe("ProductPage component", () => {
  it("renders loading state correctly", () => {
    useSelector.mockImplementation(() => ({
      loading: true,
      error: false,
      product: {},
    }));

    const wrapper = shallow(<ProductPage match={{ params: { id: "1" } }} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders error state correctly", () => {
    useSelector.mockImplementation(() => ({
      loading: false,
      error: "An error occurred",
      product: {},
    }));

    const wrapper = shallow(<ProductPage match={{ params: { id: "1" } }} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders product data correctly", () => {
    useSelector.mockImplementation(() => ({
      loading: false,
      error: false,
      product: {
        id: "1",
        name: "Product 1",
        price: 100,
        image: "image.jpg",
        rating: 4,
        numReviews: 10,
        stockKeeper: {
          _id: "123",
          stockKeeper: {
            name: "Stockkeeper",
            warehouse: "Warehouse",
          },
        },
        reviews: [],
      },
    }));

    const wrapper = shallow(<ProductPage match={{ params: { id: "1" } }} />);
    // console.log("home screen warpper >>>", wrapper);
    expect(wrapper).toMatchSnapshot();
  });
});
