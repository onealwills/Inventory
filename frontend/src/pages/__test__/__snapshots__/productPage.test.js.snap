// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductPage component renders error state correctly 1`] = `ShallowWrapper {}`;

exports[`ProductPage component renders loading state correctly 1`] = `ShallowWrapper {}`;

exports[`ProductPage component renders product data correctly 1`] = `ShallowWrapper {}`;
